pkg/cmd/cli/cmd/exporter.go

```diff
@@ -0,0 +1,216 @@
+		}
+	case *kapi.ServiceAccount:
+		serviceaccount.Strategy.PrepareForCreate(obj)
+
+	case *deployapi.DeploymentConfig:
+		// TODO: when internal refactor is completed use status reset
+		t.LatestVersion = 0
+		t.Details = nil
+		for i := range t.Triggers {
+			if p := t.Triggers[i].ImageChangeParams; p != nil {
+				p.LastTriggeredImage = ""
+			}
+		}
+	case *buildapi.BuildConfig:
+		buildconfigrest.Strategy.PrepareForCreate(obj)
+		// TODO: should be handled by prepare for create
+		t.LastVersion = 0
+		for i := range t.Triggers {
+			if p := t.Triggers[i].ImageChange; p != nil {
+				p.LastTriggeredImageID = ""
+			}
+		}
+	case *buildapi.Build:
+		buildrest.Strategy.PrepareForCreate(obj)
+		// TODO: should be handled by prepare for create
+		t.Duration = 0
+		t.Message = ""
+		t.Status = buildapi.BuildStatusNew
+		t.StartTimestamp = nil
+		t.CompletionTimestamp = nil
+		if exact {
+			return nil
+		}
+		if t.Config != nil {
+			t.Config = &kapi.ObjectReference{Name: t.Config.Name}
+		}
+	case *routeapi.Route:
+	case *imageapi.Image:
+	case *imageapi.ImageStream:
+		if exact {
+			return nil
+		}
+		// if we point to a docker image repository upstream, copy only the spec tags
+		if len(t.Spec.DockerImageRepository) > 0 {
+			t.Status = imageapi.ImageStreamStatus{}
+			break
+		}
+		// create an image stream that mirrors (each spec tag points to the remote image stream)
+		if len(t.Status.DockerImageRepository) > 0 {
+			ref, err := imageapi.ParseDockerImageReference(t.Status.DockerImageRepository)
+			if err != nil {
+				return err
+			}
+			newSpec := imageapi.ImageStreamSpec{}
+			for name, tag := range t.Status.Tags {
+				if len(tag.Items) > 0 {
+					// copy annotations
+					existing := t.Spec.Tags[name]
+					// point directly to that registry
+					ref.Tag = name
+					existing.From = &kapi.ObjectReference{
+						Kind: "DockerImage",
+						Name: ref.String(),
+					}
+					newSpec.Tags[name] = existing
+				}
+			}
+			for name, ref := range t.Spec.Tags {
+				if _, ok := t.Status.Tags[name]; ok {
+					continue
+				}
+				// TODO: potentially trim some of these
+				newSpec.Tags[name] = ref
+			}
+			t.Spec = newSpec
+			t.Status = imageapi.ImageStreamStatus{}
+			break
+		}
+
+		// otherwise, try to snapshot the most recent image as spec items
+		newSpec := imageapi.ImageStreamSpec{}
+		for name, tag := range t.Status.Tags {
+			if len(tag.Items) > 0 {
+				// copy annotations
+				existing := t.Spec.Tags[name]
+				existing.From = &kapi.ObjectReference{
+					Kind: "DockerImage",
+					Name: tag.Items[0].DockerImageReference,
+				}
+				newSpec.Tags[name] = existing
+			}
+		}
+		t.Spec = newSpec
+		t.Status = imageapi.ImageStreamStatus{}
+
+	case *imageapi.ImageStreamTag:
+		exportObjectMeta(&t.Image.ObjectMeta, exact)
+	case *imageapi.ImageStreamImage:
+		exportObjectMeta(&t.Image.ObjectMeta, exact)
+
+	default:
+		glog.V(4).Infof("No export strategy defined for objects of type %v", reflect.TypeOf(obj))
```

Comment:
We want to export things we don't recognize by default if we can clear their object meta.  It's better to be exportable with more fields (bug) than to not be exportable at all (forces a user to use another command).

----- Original Message -----
> > +					Kind: "DockerImage",
> > +					Name: tag.Items[0].DockerImageReference,
> > +				}
> > +				newSpec.Tags[name] = existing
> > +			}
> > +		}
> > +		t.Spec = newSpec
> > +		t.Status = imageapi.ImageStreamStatus{}
> > +
> > +	case *imageapi.ImageStreamTag:
> > +		exportObjectMeta(&t.Image.ObjectMeta, exact)
> > +	case *imageapi.ImageStreamImage:
> > +		exportObjectMeta(&t.Image.ObjectMeta, exact)
> > +
> > +	default:
> > +		glog.V(4).Infof("No export strategy defined for objects of type %v",
> > reflect.TypeOf(obj))
> 
> Return `ErrExportOmit` here instead?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2910/files#r31896411